### PR TITLE
V17 upgrade handler

### LINKED ISSUE #1003: v17 Upgrade Handler
- [x] Migrate host zone structs
- [x] Set slash query in progress back to false
- [x] Delete all pending queries
- [x] PFM Store key
- [x] Register community pool ICA accounts
- [x] Increase community pool tax to 2 to 5%
- [x] Change outer redemption rate bounds to get ~3months of slack
- [x] Rate limit transfer to osmosis
- [x] Remove injective rate limit
- [x] Add validator weight cap param

### PR DESCRIPTION
**Pull request by sampocs**
Closes #1003

## Context
Upgrade handler for v17 upgrade

**Note to reviewer: Please pay attention to all constants/string literals and make sure there's no typos in the chain IDs, redemption rate adjustments, rate limit thresholds, etc.**

## Changelog
* Registered community pool liquid staking ICA addresses and module addresses
* Deleted all pending slash queries
* Reset slash query in progress flag
* Increased the community pool tax
* Adjusted the redemption rate bounds
* Updated the rate limit thresholds
* Added rate limits to osmosis

### CODE REVIEW

File: app/upgrades/v17/upgrades.go
Comment (ethan-stride) [inline app/upgrades/v17/upgrades.go:173]: Why is this the only query type that we want to flush on upgrade?  Would it be a good idea to clear any slash queries since we are explicitly resetting the slashQueryInProgress flag as well?  What about more mundane things like the pool exchange rate query or queries for amounts in different ICAs, will those not have any potential interaction effects with things being modified by the upgrade handlers?

File: app/upgrades/v17/upgrades.go
Comment (ethan-stride) [inline app/upgrades/v17/upgrades.go:259]: Why only Osmosis and not any DEX?  People who are trying to exit a lot of liquidity could just use astroport or kujira fin or something instead right?

Comment (ethan-stride) [reply app/upgrades/v17/upgrades.go:259]: Would we want different limits per DEX based on amount of liquidity in the pools or would these limits be general. Is this constraint actually something which is on the Osmosis side?

**Review by ethan-stride — APPROVED** [review]
LGTM

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:173]: These are the slash queries! And the other queries you mentioned aren't live yet so they wouldn't be pending at the time that the upgrade is run

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:259]: We'll likely do a global rate limit for each sttoken in the future, across all channels

File: app/upgrades/v17/upgrades.go
Comment (riley-stride) [inline app/upgrades/v17/upgrades.go:62]: lgtm, checked against mintscan

File: app/upgrades/v17/upgrades.go
Comment (riley-stride) [reply app/upgrades/v17/upgrades.go:259]: Reopening this as I think it's a very good question. 

While we wait for the global rate limit across all channels, should we RL transfers to Neutron, Secret and Kujira too (main other DEXs)?

File: app/upgrades/v17/upgrades.go
Comment (riley-stride) [inline app/upgrades/v17/upgrades.go:82]: nit: 
```suggestion
		ctx.Logger().Info("Deleting all pending slash queries...")
```

**Review by riley-stride — CHANGES_REQUESTED** [review]
Overall looking quite good! Some small fixes we need to make, nothing major. 

Also waiting to approve as I think we're still missing (1) submit ICA to disable tokenize shares (2) bank send for shade.

> Registered community pool liquid staking ICA addresses and module addresses

lgtm

> Deleted all pending slash queries

lgtm

> Reset slash query in progress flag

lgtm

> Increased the community pool tax

LGTM. Checked it's represented as a decimal in distribution params. New value and `IncreaseCommunityPoolTax` look good.

> Adjusted the redemption rate bounds

Outer bounds lgtm. Osmosis buffer lgtm. Only question is why are we setting the inner bounds in the upgrade handler?

> Updated the rate limit thresholds

Need some changes to the percentages per PR comments. Otherwise lgtm. 

> Added rate limits to osmosis

Looks good for Osmosis, but should we RL other DEXs too?

File: x/ratelimit/keeper/hooks.go
Comment (sampocs) [reply x/ratelimit/keeper/hooks.go:18]: It was just a nit thing I noticed, I don't think it needs it's own PR.

Duration will never actually be 0 but it's a low cost way to add protection against division by 0

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:31]: They do, we probably don't need to set them! Will remove

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:50]: thanks for double checking these! Isn't POL going to bump this up very soon though?

Comment (riley-stride) [reply app/upgrades/v17/upgrades.go:50]: Yeah good point. There will be ~10m POL, of which I think 5m will be liquid staked so we'd get to ~45M. That 45M could be 50M just 10 minutes later if ATOM rallies by 10%, haha. Will defer to you here!

File: x/ratelimit/keeper/hooks.go
Comment (riley-stride) [reply x/ratelimit/keeper/hooks.go:18]: Sounds good. Resolving.

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:54]: done

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:52]: done

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:50]: ah right forgot only half is LS'd

Comment (sampocs) [reply app/upgrades/v17/upgrades.go:50]: let's do 15%! just updated

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go]: done

Comment (riley-stride) [reply app/upgrades/v17/upgrades.go]: lgtm. resolving.

File: app/upgrades/v17/upgrades.go
Comment (riley-stride) [reply app/upgrades/v17/upgrades.go]: lgtm. resolving.

File: app/upgrades/v17/upgrades.go
Comment (riley-stride) [reply app/upgrades/v17/upgrades.go]: lgtm. resolving.

File: app/upgrades/v17/upgrades.go
Comment (riley-stride) [reply app/upgrades/v17/upgrades.go]: lgtm. resolving.

**Comment by sampocs** [comment]
TODO for myself - still need to add the validator weight cap param

File: app/upgrades.go
Comment (asalzmann) [inline app/upgrades.go:221]: Is PFM the only module we're adding for v17? Need to set the store key(s)

File: app/upgrades/v17/upgrades.go
Comment (asalzmann) [inline app/upgrades/v17/upgrades.go:217]:
```suggestion
func ExecuteProp223(ctx sdk.Context, k distributionkeeper.Keeper) error {
```

for clarity

File: app/upgrades/v17/upgrades.go
Comment (asalzmann) [inline app/upgrades/v17/upgrades.go:129]: The method of setting params (resetting everything to defaults) always makes me uncomfortable.

Have we verified all params are defaults? Is there any other way to do this?

File: app/upgrades/v17/upgrades.go
Comment (asalzmann) [inline app/upgrades/v17/upgrades.go:95]:
```suggestion
		ctx.Logger().Info("Resetting slash query in progress...")
```

nit

File: app/upgrades/v17/upgrades.go
Comment (asalzmann) [inline app/upgrades/v17/upgrades.go:182]: Could a registration fail / is there any way for us to recover if one does?

File: app/upgrades/v17/upgrades.go
Comment (asalzmann) [inline app/upgrades/v17/upgrades.go:232]: nit: would this be simpler?
```
outerMin := hostZone.RedemptionRate.Mul(1-RedemptionRateOuterMinAdjustment)
```

**Review by asalzmann — COMMENTED** [review]
few small comments, otherwise lgtm!

**Comment by riley-stride** [comment]
Checking back in on this one, I think the last few tasks (besides addressing PR comments) are:

1) Add PFM store key
2) Submit ICA to disable tokenize shares
3) Bank send for shade

I've seen some of these in other PRs. Would recommend we get them all merged into here so we have a full final upgrade handler before we sign off on it.

File: app/upgrades.go
Comment (sampocs) [reply app/upgrades.go:222]: yes and added

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:225]: done

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:137]: I've verified all parameters are default and also verified with the sdk team and osmosis devs that this is the only way

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:100]: done

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:190]: Did you mean the submission itself or the full account creation?

I think the submission can only fail if the connection is closed. I think localstride should give us enough coverage here

I'm not sure what could cause the channel handshake and full account creation to fail, but I don't think it's any different than when we normally register a host

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:240]: yeah probably, but would prefer to not change it at this point

Comment (asalzmann) [reply app/upgrades/v17/upgrades.go:242]: ok

File: app/upgrades/v17/upgrades.go
Comment (asalzmann) [reply app/upgrades/v17/upgrades.go:192]: > I'm not sure what could cause the channel handshake and full account creation to fail, but I don't think it's any different than when we normally register a host

Makes sense. Can we use the standard `RestoreInterchainAccount` to restore closed channels / did you and @ethan-stride test that as part of the DYDX flow?

Comment (sampocs) [reply app/upgrades/v17/upgrades.go:192]: we modified the original tx so that it works with both host zone and trade route channels, and yeah we tested it

Comment (asalzmann) [reply app/upgrades/v17/upgrades.go:192]: Sounds good!

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:288]: we decided on doing just osmosis for now, marking this resolved

File: app/upgrades/v17/upgrades.go
Comment (asalzmann) [reply app/upgrades/v17/upgrades.go:139]: Approach makes sense - will double check all the params in the morning

**Review by riley-stride — APPROVED** [review]
I see the shade prop, the pfm store key and the ica for preventing tokenization on the hub.

Approved!

File: app/upgrades/v17/upgrades.go
Comment (sampocs) [reply app/upgrades/v17/upgrades.go:139]: double checked by @shellvish, marking this resolved

**Pull request merged**
